feat: moving the fx changes pr from documentation repo here

elnyry-sam-k opened the pull request:
to the ML specification repo on Behalf of Michael Richards for the FSPIOP SIG

elnyry-sam-k committed:
Moving the FX changes PR from documentation repo to the ML specification repo on Behalf of Michael Richards for the FSPIOP SIG

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -2798,6 +2803,422 @@ The possible states of a bulk transfer can be seen in [Figure 62](#figure-62).
+
+The arbiter of the expiry of a conversion request is the switch; or, in the case where a switch is not present, the FXP. A single point of evaluation for timeout times is required because participants in a Mojaloop system may be relatively small and informal institutions and the implementation of a system to prevent clock drift over a whole Mojaloop scheme may be challenging.
+
+No participant in a currency conversion should unilaterally time a conversion out. The requesting FSP MUST request information about the status of a conversion if it suspects that the conversion has been interrupted. The response to this request may be either a finalised state, in which case the requesting FXP MUST act according to that state, or it may be RECEIVED or PENDING. These are non-final ...
```
"The requesting FSP ...", suggesting "The FSP requesting the conversion .."

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -2798,6 +2803,422 @@ The possible states of a bulk transfer can be seen in [Figure 62](#figure-62).
+
+##### 6.13.2.9 Interledger Payment Request
+
+As part of supporting Interledger and the concrete implementation of the Interledger Payment Request (see [Section 4](#4-interledger-protocol)), the requesting FSP must attach the ILP Packet, the condition, and an expiry to the conversion. The condition and the ILP Packet are the same as those sent by the FXP in the callback of the quote; see [Section 6.5.2.3](#61223-interledger-payment-request)  ...
```
",... the requesting FSP ...", suggesting "..., the FSP requesting the conversion .."

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -2798,6 +2803,422 @@ The possible states of a bulk transfer can be seen in [Figure 62](#figure-62).
+The receiver of that transfer parses the ILP Packet to get the FXP ILP Address and routes the ILP payment by performing another transfer on the next ledger, attaching the same ILP Packet and condition and a new expiry that is less than the expiry of the incoming transfer.
+When the FXP receives the final incoming transfer to the Payee account, it extracts the ILP Packet and performs the following steps:
+1.	Validates that the Payee ILP Address in the ILP Packet corresponds to the Payee account that is the destination of the transfer.
+2.	Validates that the amount in the ILP Packet is the same as the amount of the transfer and directs the local ledger to perform a reservation of the final transfer to the Payee account (less any hidden receiver fees, see Section 5.1).
```
Is the "less any hidden receiver fees" applicable here as well?

MichaelJBRichards replied:
Yes, I think so

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -2798,6 +2803,422 @@ The possible states of a bulk transfer can be seen in [Figure 62](#figure-62).
+
+The callback **PATCH /fxTransfers/**_{ID}_ is used to inform the client of the final result of a currency conversion execution request. The _{ID}_ in the URI should contain the **commitRequestId** (see [Table 53](#table-53)) that was used for the creation of the currency conversion execution request, or the _{ID}_ that was used in the [**GET /fxTransfers/**_{ID}_](#61331-get-fxtransfers). See [Ta ...
+
+Specifically, it is used by the finalising entity (either the optional switch or the payee FSP) to inform an FXP of the finalised status of a currency conversion where the currency conversion is dependent on the finalised status of a determining transfer. The finalising entity will send a **PATCH /fxTransfers/_{ID}_** message to the FXP to signal the final status of the associated payment, so tha ...
```
payee FSP -> Payee FSP

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -2798,6 +2803,422 @@ The possible states of a bulk transfer can be seen in [Figure 62](#figure-62).
+
+Alternative URI: N/A
+
+Logical API service: **Return Currency conversion Execution Information Error**
```
"**Return Currency conversion Execution Information Error**" -> "**Return Currency Conversion Execution Information Error**" (capital C in "Conversion")

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3879,17 +4366,20 @@ This section describes complex types used by the API.
 | **amount** | 1 | Money | Transaction amount to be sent. |
+| **payeeReceiveAmount** | 0..1 | Money | The amount that the beneficiary will receive |
 | **transactionType** | 1 | TransactionType | Type of the transaction. |
+|**converter** | 0..1 | CurrencyConverter | **PAYER** if the payer DFSP intends to perform currency conversion; **PAYEE** if the payer DFSP wants the payee DFSP to perform currency conversion. If absent and the transfer requires currency conversion, then the converting institution will be inferred from the currency in which the quotation is requested: if the quotation is requested in the source c ...
```
payer DFSP -> Payer FSP, payer DFSP -> Payer FSP

henrka replied:
"| Used by the debtor party if it wants to share information about the currency conversion it proposes to make; or if it is required by scheme rules to share this information. This object contains the amount of the transfer in the source and target currencies, but does not identify the FXP being used. |"
Seems like a copy-paste error here, same is in line below.

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3879,17 +4366,20 @@ This section describes complex types used by the API.
+| **payeeReceiveAmount** | 0..1 | Money | The amount that the beneficiary will receive |
 | **transactionType** | 1 | TransactionType | Type of the transaction. |
+|**converter** | 0..1 | CurrencyConverter | **PAYER** if the payer DFSP intends to perform currency conversion; **PAYEE** if the payer DFSP wants the payee DFSP to perform currency conversion. If absent and the transfer requires currency conversion, then the converting institution will be inferred from the currency in which the quotation is requested: if the quotation is requested in the source c ...
+| **conversionRate** | 0..1 | FxRate | Used by the debtor party if it wants to share information about the currency conversion it proposes to make; or if it is required by scheme rules to share this information. This object contains the amount of the transfer in the source and target currencies, but does not identify the FXP being used. |
```
"Used by the debtor party if it wants to share information ...", debtor party is never used in this spec, suggesting "Used by the Payer FSP if it wants to share information ..."

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3960,17 +4450,30 @@ This section contains the enumerations that are used by the API.
+
+| **Name** | **Description** |
+| --- | --- |
+| **PAYER** | Currency conversion should be performed by the payer DFSP |
```
payer DFSP -> Payer FSP

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3960,17 +4450,30 @@ This section contains the enumerations that are used by the API.
+| **Name** | **Description** |
+| --- | --- |
+| **PAYER** | Currency conversion should be performed by the payer DFSP |
+| **PAYEE** | Currency conversion shold be performed by the payee DFSP. |
```
payee DFSP -> Payee FSP

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -1978,6 +1981,8 @@ Callback and data model information for **POST /quotes**:
 | **payer** | 1 | Party | Information about the Payer in the proposed financial transaction. |
 | **amountType** | 1 | AmountType |**SEND** for send amount, **RECEIVE** for receive amount. |
 | **amount** | 1 | Money | Depending on **amountType**:<br>If **SEND**: The amount the Payer would like to send; that is, the amount that should be withdrawn from the Payer account including any fees. The amount is updated by each participating entity in the transaction.<br>If **RECEIVE**: The amount the Payee should receive; that is, the amount that should be sent to the receiver exclusive any f ...
+|**converter** | 0..1 | CurrencyConverter | **PAYER** if the payer DFSP intends to perform currency conversion; **PAYEE** if the payer DFSP wants the payee DFSP to perform currency conversion. If absent and the transfer requires currency conversion, then the converting institution will be inferred from the currency in which the quotation is requested: if the quotation is requested in the source c ...
```
payer DFSP -> Payer FSP, payee DFSP -> Payee FSP

henrka replied:
"| Used by the debtor party if it wants to share information about the currency conversion it proposes to make; or if it is required by scheme rules to share this information. This object contains the amount of the transfer in the source and target currencies, but does not identify the FXP being used. |"
This seems like a copy-paste error, the same information is in line below as well.

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3209,485 +3630,550 @@ This section defines elements types used by the API.
+| **determiningTransferId** | 0..1 | CorrelationId | The transaction ID of the transfer on whose success this currency conversion depends. If this is a bulk currency conversion which is not dependent on a transfer, then this field should be omitted. |
+| **counterPartyFsp** | 1 | FspId | The ID of the FXP performing the conversion. |
+| **amountType** | 1 | AmountType | This is the AmountType for the base transaction, as described in Section 7.3.1 above.<ln>
+If it is set to SEND, then any charges levied by the FXP as part of the transaction will be deducted by the FXP from the amount shown for the target party in the conversion.<ln>
```
I don't think this should be on it's own line.. See how it is done for IndividualQuote.

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3209,485 +3630,550 @@ This section defines elements types used by the API.
+| **counterPartyFsp** | 1 | FspId | The ID of the FXP performing the conversion. |
+| **amountType** | 1 | AmountType | This is the AmountType for the base transaction, as described in Section 7.3.1 above.<ln>
+If it is set to SEND, then any charges levied by the FXP as part of the transaction will be deducted by the FXP from the amount shown for the target party in the conversion.<ln>
+If it is set to RECEIVE, then any charges levied by the FXP as part of the transaction will be added by the FXP to the amount shown for the source party in the conversion. |
```
I don't think this should be on it's own line.. See how it is done for IndividualQuote.

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3209,485 +3630,550 @@ This section defines elements types used by the API.
+| **amountType** | 1 | AmountType | This is the AmountType for the base transaction, as described in Section 7.3.1 above.<ln>
+If it is set to SEND, then any charges levied by the FXP as part of the transaction will be deducted by the FXP from the amount shown for the target party in the conversion.<ln>
+If it is set to RECEIVE, then any charges levied by the FXP as part of the transaction will be added by the FXP to the amount shown for the source party in the conversion. |
+|initiatingFsP | 1 | fspId | The id of the participant who is requesting a currency conversion. |
```
fspId -> FspId (types begin with Capital)

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3209,485 +3630,550 @@ This section defines elements types used by the API.
-**Table 84 -- Complex type ExtensionList**
+**Table 98 -- Complex type ExtensionList**
+
+#### 7.4.5 Extension
```
Shouldn't this be FxMoney?

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3209,485 +3630,550 @@ This section defines elements types used by the API.
+
+**Table 99 -- Complex type Extension**
 
 #### 7.4.5 IndividualQuote
```
Seems to be double 7.4.5 now..

henrka reviewed (commented):
Please find my comments after a first read of the actual changes.

MichaelJBRichards committed:
Update API Definition v2.0-alpha.md

Addressing PR comments

MichaelJBRichards commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -2798,6 +2803,422 @@ The possible states of a bulk transfer can be seen in [Figure 62](#figure-62).
+
+| **Name** | **Cardinality** | **Type** | **Description** |
+| --- | --- | --- | --- |
+| **providers** | 0..16 | FspId | The FSP Id(s) of the participant(s) who offer currency conversion services for the currency corridor defined in the URI. |
```
Increased it to 32...

MichaelJBRichards commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -2798,6 +2803,422 @@ The possible states of a bulk transfer can be seen in [Figure 62](#figure-62).
+When the Payee FSP presents the fulfilment to the common ledger, the transfer is committed in the common ledger. At the same time, the Payer FSP is notified that the transfer has been committed along with the fulfilment.
+
+Mojaloop supports two forms of currency conversion. The first is stand-alone conversion, in which the converted funds are returned to the requester's account. To use this functionality, the requesting DFSP must maintain a ledger in the target currency. The second form is Payment-versus-Payment (PvP) conversion. In this form, execution of the conversion is dependent on successful completion of a l ...
+7.4.3](#743-conversion) below).
```
Not sure that I understand this comment

MichaelJBRichards commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -2798,6 +2803,422 @@ The possible states of a bulk transfer can be seen in [Figure 62](#figure-62).
+
+##### 6.13.2.2 Expired Quote
+
+If an FSP issues an instruction to execute a conversion using a conversion quotation which has expired, then the FXP MAY reject the request.
```
Hmmm. My feeling is that the FXP should be able to accept the request if it wants to (for instance, if the request has only just expired); but is under no obligation to...

henrka replied:
Wouldn't that lead to potential inconsistencies? At least I would expect an expired quote to be rejected, now as a client I have no idea how the FXP will behave.

MichaelJBRichards commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -2798,6 +2803,422 @@ The possible states of a bulk transfer can be seen in [Figure 62](#figure-62).
+
+The requesting FSP must always set a transfer expiry time to allow for use cases in which a swift completion or failure is needed. If the use case does not require a swift completion, a longer expiry time can be set.
+
+The arbiter of the expiry of a conversion request is the switch; or, in the case where a switch is not present, the FXP. A single point of evaluation for timeout times is required because participants in a Mojaloop system may be relatively small and informal institutions and the implementation of a system to prevent clock drift over a whole Mojaloop scheme may be challenging.
```
See what you think...

MichaelJBRichards committed:
Update API Definition v2.0-alpha.md

Responses to Henrik's comments

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -24,7 +24,7 @@ specified types of information.
 |**1.1**|2020-05-19|1. This version contains a new option for a Payee FSP to request a commit notification from the Switch. The Switch should then send out the commit notification using the new request **PATCH /transfers/**_{ID}_. The option to use commit notification replaces the previous option of using the ”Optional Additional Clearing Check”. The section describing this has been replaced with ...
 |**1.1.1**|2021-09-22|This document version only adds information about optional HTTP headers regarding tracing support in [Table 2](#table-2), see _Distributed Tracing Support for OpenAPI Interoperability_ for more information. There are no changes in any resources as part of this version.|
-|**2.0**|2021-10-26|- The element named **initiatorType** has been removed from the complex type [**TransactionType**](#7418-transactiontype). This information is now supposed to be carried as part of the complex type [**Party**](#7411-party) instead, to be able to have more detail of the transaction by allowing to state the type of both the Payer and Payee. Section [5.3 Mapping of Use Cases to T ...
+|**2.0**|2021-10-26|- The element named **initiatorType** has been removed from the complex type [**TransactionType**](#7418-transactiontype). This information is now supposed to be carried as part of the complex type [**Party**](#7411-party) instead, to be able to have more detail of the transaction by allowing to state the type of both the Payer and Payee. Section [5.3 Mapping of Use Cases to T ...
```
Can you add a link to https://github.com/mojaloop/mojaloop-specification/issues/89, so that it is possible to trace the origin of the change?

henrka replied:
@MichaelJBRichards, please add a link to #89 in the following: "-New endpoints and data structures have been added to support currency conversion.", so that you can understand the origin of the change.

MichaelJBRichards replied:
Done

MichaelJBRichards replied:
Done

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -2822,23 +2822,23 @@ Table 45 contains a description of each different version of the **/services** r
 ##### 6.11.2.1 Currency Conversion Services 
 
-A request for participants in a Mojaloop system who support currency conversion services (Foreign Exchange Providers or FXPs) is made using the syntax **services/FXP**
+A request for FSPs who support currency conversion services (that is, FSPs who have the rolw of Foreign Exchange Providers or FXPs) is made using the syntax **services/FXP**
```
".. (that is, FSPs who have the rolw of Foreign Exchange Providers ..." rolw -> role

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -2904,25 +2904,25 @@ A currency conversion quote is irrevocable; it cannot be changed after it has be
+Currency conversion is a process that takes place between two FSPs: the requesting FSP and the FXP. Although the success or failure of a currency conversion may depend on the outcome of a related payment, the conversion itself is a transaction between the two FSPs. The FSP is asking the FXP to provide cover for the payment it intends to make. The conversion does not, therefore, require KYC inform ...
 
-Either the payer DFSP or the payee DFSP may request currency conversion. If the payer DFSP requests conversion, then the transfer itself will be denominated in the target currency and the payee DFSP need know nothing about the fact that the conversion was performed: as far as it is concerned, the transfer is in the currency they expect. Conversely, if the payee DFSP is undertaking the conversion, ...
+Either the payer FSP or the payee FSP may request currency conversion. If the payer FSP requests conversion, then the transfer itself will be denominated in the target currency and the payee FSP need know nothing about the fact that the conversion was performed: as far as it is concerned, the transfer is in the currency they expect. Conversely, if the payee FSP is undertaking the conversion, then ...
```
"payer FSP" to "Payer FSP", "payee FSP" to "Payee FSP" to align with elsewhere in the document..

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -2933,22 +2933,20 @@ Either the payer DFSP or the payee DFSP may request currency conversion. If the
 ##### 6.12.2.1 Conversion Quote Expiry Details
 
-In the same way as with an ordinary quotation (see [Section 6.5.2.1](#6521-quote-expiry-details) above), the requesting DFSP should include an expiry time for the quotation.
+In the same way as with an ordinary quotation (see [Section 6.5.2.1](#6521-quote-expiry-details) above), the FSP requesting the quotation should include an expiry time for the quotation.
```
Suggesting to remove "above", as using "above" might increase the required maintenance of the document (a move might place the section below instead)..

elnyry-sam-k replied:
changed.

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -2997,8 +2995,8 @@ The callback **PUT /fxQuotes/**_{ID}_ is used to inform the client of a requeste
 | **condition** | 0..1 | ilpCondition | The ILP condition for the conversion. |
-| **ilpPacket** | 0..1 | ilpPacket | The ILP Packet containing the amount delivered to the Payee, the ILP Address of the Payee and any other end-to-end data. |
-| **conversionTerms** | 1 | Conversion | The terms under which the FXP will undertake the currency conversion proposed by the requester.  |
+| **ilpPacket** | 0..1 | ilpPacket | The ILP Packet containing the amount delivered to the requestingFSP, the ILP Address of the requesting FSP and any other end-to-end data. |
```
I think you used "client" instead of "requesting FSP" earlier.. Below "the FSP that requested the conversion" is used.

MichaelJBRichards replied:
I think it is the Payee FSP, on reflection...

elnyry-sam-k replied:
Text has been updated so resolving this

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3097,7 +3092,7 @@ If the conversion was a PvP request dependent on a payment, then that conversion
 ##### 6.13.2.9 Interledger Payment Request
 
-As part of supporting Interledger and the concrete implementation of the Interledger Payment Request (see [Section 4](#4-interledger-protocol)), the requesting FSP must attach the ILP Packet, the condition, and an expiry to the conversion. The condition and the ILP Packet are the same as those sent by the FXP in the callback of the quote; see [Section 6.5.2.3](#61223-interledger-payment-request)  ...
+As part of supporting Interledger and the concrete implementation of the Interledger Payment Request (see [Section 4](#4-interledger-protocol)), the FSP requsting the conversion must attach the ILP Packet, the condition, and an expiry to the conversion. The condition and the ILP Packet are the same as those sent by the FXP in the callback of the quote; see [Section 6.5.2.3](#61223-interledger-pay ...
```
"requsting" -> "requesting"

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3075,11 +3070,11 @@ If an FSP issues an instruction to execute a conversion using a conversion quota
+The FSP requesting the conversion must always set a transfer expiry time to allow for use cases in which a swift completion or failure is needed. If the use case does not require a swift completion, a longer expiry time can be set.
 
-The arbiter of the expiry of a conversion request is the switch; or, in the case where a switch is not present, the FXP. A single point of evaluation for timeout times is required because participants in a Mojaloop system may be relatively small and informal institutions and the implementation of a system to prevent clock drift over a whole Mojaloop scheme may be challenging.
+The arbiter of the expiry of a conversion request is the switch; or, in the case where a switch is not present, the FXP. A single point of evaluation for timeout times is required because the communicating participants may be relatively small and informal institutions and the implementation of a system to prevent clock drift among all participants may be challenging.
```
Wouldn't it be easier to understand if "FSPs" is used instead of "communicating participants"?

MichaelJBRichards committed:
Update API Definition v2.0-alpha.md

Responses to Henrik's review 20240715

henrka reviewed (commented):
Please also fix the Introduction and Table 6, 7, and 8 as discussed on chat.

MichaelJBRichards committed:
Update API Definition v2.0-alpha.md

Responding to comments from Henrik Karlsson.

MichaelJBRichards committed:
Update API Definition v2.0-alpha.md

Added links to specification sections from introduction

MichaelJBRichards committed:
Update API Definition v2.0-alpha.md

Changes to support ULIDs as unique identifiers as well as UUIDs

elnyry-sam-k committed:
Merge branch 'master' into feat/fx-changes

MichaelJBRichards committed:
Update API Definition v2.0-alpha.md

Add new header parameter and changes to accept and content-type parameter syntax.

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -24,7 +24,7 @@ specified types of information.
 |**1.1**|2020-05-19|1. This version contains a new option for a Payee FSP to request a commit notification from the Switch. The Switch should then send out the commit notification using the new request **PATCH /transfers/**_{ID}_. The option to use commit notification replaces the previous option of using the ”Optional Additional Clearing Check”. The section describing this has been replaced with ...
 |**1.1.1**|2021-09-22|This document version only adds information about optional HTTP headers regarding tracing support in [Table 2](#table-2), see _Distributed Tracing Support for OpenAPI Interoperability_ for more information. There are no changes in any resources as part of this version.|
-|**2.0**|2021-10-26|- The element named **initiatorType** has been removed from the complex type [**TransactionType**](#7418-transactiontype). This information is now supposed to be carried as part of the complex type [**Party**](#7411-party) instead, to be able to have more detail of the transaction by allowing to state the type of both the Payer and Payee. Section [5.3 Mapping of Use Cases to T ...
+|**2.0**|2021-10-26|- The element named **initiatorType** has been removed from the complex type [**TransactionType**](#7418-transactiontype). This information is now supposed to be carried as part of the complex type [**Party**](#7411-party) instead, to be able to have more detail of the transaction by allowing to state the type of both the Payer and Payee. Section [5.3 Mapping of Use Cases to T ...
```
Please add a link to #120 in the following: "- The definition of a correlation ID has been amended to support the use of ULIDs as well as UUIDs as globally unique identifiers."

elnyry-sam-k replied:
added

ridhiarora26 replied:
It seems issue points to #130 and not #120

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -24,7 +24,7 @@ specified types of information.
 |**1.1**|2020-05-19|1. This version contains a new option for a Payee FSP to request a commit notification from the Switch. The Switch should then send out the commit notification using the new request **PATCH /transfers/**_{ID}_. The option to use commit notification replaces the previous option of using the ”Optional Additional Clearing Check”. The section describing this has been replaced with ...
 |**1.1.1**|2021-09-22|This document version only adds information about optional HTTP headers regarding tracing support in [Table 2](#table-2), see _Distributed Tracing Support for OpenAPI Interoperability_ for more information. There are no changes in any resources as part of this version.|
-|**2.0**|2021-10-26|- The element named **initiatorType** has been removed from the complex type [**TransactionType**](#7418-transactiontype). This information is now supposed to be carried as part of the complex type [**Party**](#7411-party) instead, to be able to have more detail of the transaction by allowing to state the type of both the Payer and Payee. Section [5.3 Mapping of Use Cases to T ...
+|**2.0**|2021-10-26|- The element named **initiatorType** has been removed from the complex type [**TransactionType**](#7418-transactiontype). This information is now supposed to be carried as part of the complex type [**Party**](#7411-party) instead, to be able to have more detail of the transaction by allowing to state the type of both the Payer and Payee. Section [5.3 Mapping of Use Cases to T ...
```
Strange to have the link in "A new element" in the following: "- [A new element](https://github.com/mojaloop/mojaloop-specification/issues/130), **FSPIOP-Proxy**, has been added to the set of message header fields." Suggesting putting the link at the end to keep same structure as other bullets.

elnyry-sam-k replied:
updated to have the link at the end

henrka commented:
Not part of this change, but suggesting to update the following in Section "2. Introduction":
> Version 1.0 of the API is designed to be used within a country or region, international remittance that requires foreign exchange is not supported. This version also contains basic support for the Interledger Protocol, which will in future versions of the API be used for supporting foreign exchange and multi-hop financial transactions.

To (removing part with "future versions":
> Version 1.0 of the API is designed to be used within a country or region, international remittance that requires foreign exchange is not supported. This version also contains basic support for the Interledger Protocol.

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
````diff
@@ -445,7 +446,7 @@ See below for an example of a simplified HTTP request which only includes an **A
 ```
 POST /service HTTP/1.1
-Accept: application/vnd.interoperability.{resource}+json;version=1,
+Accept: application/vnd.interoperability.{messagetype}.{resource}+json;version=1,
````
Hmm I think the old structure was supposed to be kept as is for FSPIOP? @PaulGregoryBaker to confirm..

elnyry-sam-k replied:
hi @henrka 

Accept: application/vnd.interoperability.{messagetype}.{resource}+json;version=1, 

is the proposed new structure

henrka replied:
Why not include the earlier version as default to limit changes for existing implementers? Meaning if you use "Accept: application/vnd.interoperability.{resource}+json;version=1" then fspiop is used, then you can use "Accept: application/vnd.interoperability.{messagetype}.{resource}+json;version=1" if you would like to set a specific type.

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -459,7 +460,7 @@ Regarding the example in [Listing 3](#listing-3):
 - The **_POST /service_** should be changed to any HTTP method and related service or resource that is supported by the API (see [Table 6](#table-6)).
 - The **Accept** header field is used to indicate the API resource version the client would like to use. If several versions are supported by the client, more than one version can be requested separated by a comma (**,**) as in the example above.
-  - The application type is always **application/vnd.interoperability.**_{resource}_, where _{resource}_ is the actual resource (for example, **participants** or **quotes**).
+  - The application type is always **application/vnd.interoperability**._{messagetype}_._{resource}_, where {messagetype} is the format used for the message (for example, **fspiop** or **iso20022**), and {resource} is the actual resource (for example, **participants** or **quotes**).
```
Hmm I think the old structure was supposed to be kept as is for FSPIOP? @PaulGregoryBaker to confirm..

elnyry-sam-k replied:
hi @henrka

Accept: application/vnd.interoperability.{messagetype}.{resource}+json;version=1,

is the proposed new structure

henrka replied:
Why not include the earlier version as default to limit changes for existing implementers? Meaning if you use "Accept: application/vnd.interoperability.{resource}+json;version=1" then fspiop is used, then you can use "Accept: application/vnd.interoperability.{messagetype}.{resource}+json;version=1" if you would like to set a specific type.

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
````diff
@@ -471,7 +472,7 @@ If the server supports the API resource version requested by the client in the A
 
 ```
-Content-Type: application/vnd.interoperability.resource+json;version=1.0
+Content-Type: application/vnd.interoperability.messagetype.resource+json;version=1.0
````
Hmm I think the old structure was supposed to be kept as is for FSPIOP? @PaulGregoryBaker to confirm..

elnyry-sam-k replied:
hi @henrka

Accept: application/vnd.interoperability.{messagetype}.{resource}+json;version=1,

is the proposed new structure based on what was implemented; please review

henrka replied:
Why not include the earlier version as default to limit changes for existing implementers? Meaning if you use "Accept: application/vnd.interoperability.{resource}+json;version=1" then fspiop is used, then you can use "Accept: application/vnd.interoperability.{messagetype}.{resource}+json;version=1" if you would like to set a specific type.

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -505,6 +506,11 @@ Along with HTTP status 406, the supported versions should be listed as part of t
 **Listing 5 -- Example error message when server does not support the requested version**
 
+
+#### 3.3.4.4 Message type determination
```
Why do we even need to know about ISO version in FSPIOP API?

MichaelJBRichards replied:
This is for disambiguation purposes. We do not expect people to know about the other protocol; but since there is a possibility of information being left in one of the definitions but not the other, we need to be explicit about interpretation.

henrka commented:
Table 8, the following is on its own row which looks incorrect:

> For consistency, the data model for the POST /participants/{Type}/{ID} and POST /participants/{Type}/{ID}/{SubId} calls in Table 10 has been updated to include the optional ExtensionList element as well.

henrka commented:
Table 7 should include the change on correlation IDs. This is a breaking change for all version 1.x, so all versions need to be bumped to 2.0 to capture the breaking change.

The same information update is then required for Tables 8, 15, 18, 22, 26, 29, 34, 37, and 41.

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -97,15 +99,16 @@ The API is based on the REST (REpresentational State Transfer<sup>1</sup>) archi
    - The client can use the common ID in an HTTP **GET** request directly if it does not receive a callback from the server (see [Section 3.2.2](#322-http-methods) for more information).
 
-  To keep the common IDs unique, each common ID is defined as a UUID (Universally Unique IDentifier<sup>2</sup> (UUID). To further guarantee uniqueness, it is recommended that a server should separate each client FSP's IDs by mapping the FSP ID and the object ID together. If a server still receives a non-unique common ID during an HTTP **POST** request (see [Section 3.2.2](#322-http-methods) for  ...
+  To keep the common IDs unique, each common ID must be generated using a technique which will create an identifier which is reliably globally unique. The Mojaloop specification does not require a particular technique, but two methods are currently supported:  lowercase UUID<sup>2</sup> and uppercase ULID <sup>3</sup>. To further guarantee uniqueness, it is recommended that a server should separa ...
```
> The Mojaloop specification does not require a particular technique, ..

This is the only instance of "Mojaloop specification" in the document, and it is not **the** "Mojaloop specification". Recommending to change from "The Mojaloop specification" to just "This specification".

elnyry-sam-k replied:
updated

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -204,17 +210,18 @@ The API supports a maximum size of 65536 bytes (64 Kilobytes) in the HTTP header
+|**Content-Length**|**3495**|0..1|The **Content-Length**<sup>12</sup> header field indicates the anticipated size of the payload body. Only sent if there is a body.<br>**Note**: The API supports a maximum size of 5242880 bytes (5 Megabytes).<br>|
+|**Content-Type**|**application/vnd.interoperability.resource+json;version=1.0**|1|The **Content-Type**<sup>13</sup> header indicates the specific version of the API used to send the payload body. See [Section 3.3.4.2](#3342-acceptable-version-requested-by-client) for more information.|
+|**Date**|**Tue, 15 Nov 1994 08:12:31 GMT**|1|The **Date**<sup>14</sup> header field indicates the date when the request was sent.|
+|**X- Forwarded- For**|**X-Forwarded-For: 192.168.0.4, 136.225.27.13**|1..0|The **X-Forwarded-For**<sup>15</sup> header field is an unofficially accepted standard used to indicate the originating client IP address for informational purposes, as a request might pass multiple proxies, firewalls, and so on. Multiple **X-Forwarded-For** values as in the example shown here should be expected and suppo ...
```
Seems like there has been an error in the below while moving from Word version to markdown. "RFC 723915" should be "RFC 7239", the "15" should be the reference.

> An alternative to **X-Forwarded-For** is defined in RFC 723915

elnyry-sam-k replied:
updated

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -1371,7 +1389,9 @@ On a high level, the API can be used to perform the following actions:
 |/bulkQuotes|2.0|The data model for a [**Party**](#7411-party) is updated to add an optional type of the [**Party**](#7411-party) in the element [**PartyType**](#7330-partytype)). This change is based on the Change Request: https://github.com/mojaloop/mojaloop-specification/issues/25. It is also updated to add two new [**PartyIDType**](#Table-103)s based on the Change Request: https://github.com/ ...
 |/bulkTransfers|2.0|The data model for a [**Party**](#7411-party) is updated to add an optional type of the [**Party**](#7411-party) in the element [**PartyType**](#7330-partytype)). This change is based on the Change Request: https://github.com/mojaloop/mojaloop-specification/issues/25. It is also updated to add two new [**PartyIDType**](#Table-103)s based on the Change Request: https://github.c ...
-
+|/services|1.0|Initial version
```
Wouldn't it be more consistent and easy to understand if this was also 2.0, like all other versions are being bumped to? Please also add a final | to end the row.

elnyry-sam-k replied:
updated to 2.0

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -1371,7 +1389,9 @@ On a high level, the API can be used to perform the following actions:
 |/bulkTransfers|2.0|The data model for a [**Party**](#7411-party) is updated to add an optional type of the [**Party**](#7411-party) in the element [**PartyType**](#7330-partytype)). This change is based on the Change Request: https://github.com/mojaloop/mojaloop-specification/issues/25. It is also updated to add two new [**PartyIDType**](#Table-103)s based on the Change Request: https://github.c ...
-
+|/services|1.0|Initial version
+|/fxQuotes|1.0|Initial version
```
Wouldn't it be more consistent and easy to understand if this was also 2.0, like all other versions are being bumped to? Please also add a final | to end the row.

elnyry-sam-k replied:
updated to 2.0

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -1371,7 +1389,9 @@ On a high level, the API can be used to perform the following actions:
-
+|/services|1.0|Initial version
+|/fxQuotes|1.0|Initial version
+|/fxTransfers|1.0|Initial version
```
Wouldn't it be more consistent and easy to understand if this was also 2.0, like all other versions are being bumped to? Please also add a final | to end the row (this is what is causing the "Table 7 – Current resource versions" to be part of the actual table).

elnyry-sam-k replied:
updated to 2.0

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3133,27 +3566,29 @@ Two examples of the **Date** type appear below:
 **1987-08-05**
 
-#### 7.2.16 UUID
+#### 7.2.16 Uniqueidentifier
```
**Uniqueidentifier** -> **UniqueIdentifier** (capital I in Identifier)

elnyry-sam-k replied:
updated

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3209,485 +3644,548 @@ This section defines elements types used by the API.
+#### 7.3.12 ErrorCode
 
-[Table 56](#table-56) contains the data model for the element **ErrorCode**.
+[Table ]69(#table-69) contains the data model for the element **ErrorCode**.
```
> [Table ]69(#table-69)

Should be [Table 69](#table-69) (misplaced "]")

elnyry-sam-k replied:
updated

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3209,485 +3644,548 @@ This section defines elements types used by the API.
+
+| **Name** | **Cardinality** | **Format** | **Description** |
+| --- | --- | --- | --- |
+| **chargeType** | 1 | String(1..32) | A descripition of the charge that is being levied |
```
descripition -> description

elnyry-sam-k replied:
updated

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3209,485 +3644,548 @@ This section defines elements types used by the API.
+| **Name** | **Cardinality** | **Format** | **Description** |
+| --- | --- | --- | --- |
+| **currency** | 1 | Currency | The currency of the amount, as described in Section 7.3.9 above.
+ |
```
This should end the row above instead of being on its own line.

henrka commented on fspiop-api/documents/API Definition v2.0-alpha.md:
```diff
@@ -3209,485 +3644,548 @@ This section defines elements types used by the API.
+
+| **Name** | **Cardinality** | **Format** | **Description** |
+| --- | --- | --- | --- |
+| **currency** | 1 | Currency | The currency of the amount, as described in Section 7.3.9 above.
```
Remove "above", as this makes it hard to maintain the document.

elnyry-sam-k replied:
removed

henrka reviewed (commented):
Review of the updated document, see also generic comments which are not directly related to the updated parts.

elnyry-sam-k committed:
changes based on review and feedback by Henrik
